### PR TITLE
Biteau catalog new

### PR DESCRIPTION
**Pull request by pdeiml**
Hi Christoph,

please check the calculation of the energy and of the flux which are in line 58f and line 63f in Add_Biteau.py, respectively.

Current status:
All data sets in Biteau's catalog are added in corresponding ecsv-files in /input/data/<year>/<reference_id>

TODO:
- [x] Error handling in make.py checks
- [x] Add all Biteau-references to /input/gammacat/gamma_cat_dataset.yaml
- [x] Add all Biteau-references to /input/sources/tev-<SOURCE_ID>.yaml

### CODE REVIEW

**Comment by cdeil** [comment]
@pdeiml - Some comments:

- Please don't delete `biteau.py` here, just leave as-is. I'd prefer to keep it if I have to come back to this in the future.
- Please commit the generated SED files in `input/data` and add them here. The first thing I would like to check / review is whether the SED you have is consistent to what I computed https://gist.github.com/cdeil/2942d2f0241d7ff8d25db8008e0fda03 when reviewing #120 .

File: other_data_collections/2015ApJ...812...60B/Add_Biteau.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+    def Rename_Experiment(self, experiment):
+        return self.fBiteauExperiments[experiment]
+
+# Calculate_Energy transforms the 'freq'-column in Biteau's catalog into an energy-column 
+    def Calculate_Energy(self, frequency):
+        energy = (h*frequency*1/u.second)/e.si * 1E-9* u.teraelectronvolt/u.joule*u.coulomb
+        return energy
+
+# Calculate_dnde transforms the 'e2dnde'-column in Biteau's catalog into an dnde-column
+    def Calculate_dnde(self, energy, e2dnde):
```
Comment (cdeil) [inline other_data_collections/2015ApJ...812...60B/Add_Biteau.py]: The functions you have to compute `energy` and `dnde` can be improved:
* Make them standalone functions above your class (methods that take `self` as argument but then don't use it aren't nice)
* I don't really understand why the code you have to compute these is such a long line, what I have in https://nbviewer.jupyter.org/gist/cdeil/2942d2f0241d7ff8d25db8008e0fda03 is shorter / more readable, no?
In any case, there's still "magic constants" like 1e9 and 1e-16 in the formula, that shouldn't be necessary if you use quantities. And the line for `dnde` is too long, I have to scroll back and forth on Github to review it. So bottom line: please try to make it more readable, using the code in the notebook I linked to as example.

File: other_data_collections/2015ApJ...812...60B/Add_Biteau.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+        biteau_sources = dict()
+        print('Reading {}'.format(filename))
+        data = ascii.read(filename)
+        for i in range(0, len(data)):
+            biteau_sources[data[i][0]] = data[i][1]
+        return biteau_sources
+
+# Load_Experiments read Biteau_Experiments.txt in which all sources which are mentioned in
+# Biteau's catalog and the corresponding telescope from /input/schemas are stored
+    def Load_Experiments(self, filename):
```
Comment (cdeil) [inline other_data_collections/2015ApJ...812...60B/Add_Biteau.py]: The code to read your ascii files and make a dict for name mapping can be simplified to this:
```
from astropy.table import Table
table = Table.read('Biteau_Experiments.txt', format='ascii')
dict(zip(table['experiment(Biteau)'], table['telescope']))
```

File: other_data_collections/2015ApJ...812...60B/Add_Biteau.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+        biteau_experiments = dict()
+        print('Reading {}'.format(filename))
+        data = ascii.read(filename)
+        for i in range(0, len(data)):
+            biteau_experiments[data[i][0]] = data[i][1]
+        return biteau_experiments
+
+# Rename_Experiment renames all source names in Biteau's catalog into the corresponding name
+# in /input/schemas
+    def Rename_Experiment(self, experiment):
```
Comment (cdeil) [inline other_data_collections/2015ApJ...812...60B/Add_Biteau.py]: This one-line method isn't useful -> suggest to remove and just use the dict directly.

File: other_data_collections/2015ApJ...812...60B/Add_Biteau.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+        return energy
+
+# Calculate_dnde transforms the 'e2dnde'-column in Biteau's catalog into an dnde-column
+    def Calculate_dnde(self, energy, e2dnde):
+        dnde = 1/((energy*u.teraelectronvolt)**2)*e2dnde*1/e.si*1E-16*u.erg*1/((u.centimeter)**2)/u.second*u.teraelectronvolt/u.erg
+        return dnde
+
+# Create_Subtables splits up Biteau's catalog into the single data sets and stores these sets 
+# in a list which contains astropy tables
+    def Create_Subtables(self):
```
Comment (cdeil) [inline other_data_collections/2015ApJ...812...60B/Add_Biteau.py]: I think your code to make the subtables is much longer / complex than it needs to be.
Suggest to make a mask with the row subset selection and then index into the table to get the sub-table with that mask, like I did here: https://nbviewer.jupyter.org/gist/cdeil/2942d2f0241d7ff8d25db8008e0fda03
Should just be ~ 5 lines, no?

**Comment by pdeiml** [comment]
I checked the spectral fluxes of source 49 (https://gist.github.com/cdeil/2942d2f0241d7ff8d25db8008e0fda03) and Add_Biteau.py calculates it correctly.
If you agree with everything up to know, I will work off the enumeration in my first comment.

**Comment by cdeil** [comment]
OK, great! Let me know if you have more questions or when it's ready for final review.

**Comment by pdeiml** [comment]
In principle this PR is finished.

"make.py checks" still produces some errors:
1) The ecsv-files which are generated from Biteau's catalog do have the meta-key "mjd" which is currently not allowed or not known.
Should I add this key to 
https://github.com/gammapy/gamma-cat/blob/master/input/schemas/dataset_source_info.schema.yaml ??
2) The new references are not listed in 
https://github.com/gammapy/gamma-cat/blob/master/input/gammacat/gamma_cat_dataset.yaml
Should I add the references?

**Comment by cdeil** [comment]
The MJD meta key for the SED files should be added to the list here:
https://github.com/gammapy/gamma-cat/blob/master/gammacat/sed.py#L47

Yes, if you list the new references in `gamma_cat_dataset.yaml` that would be good, that will mean they are used for the catalog.

**Comment by pdeiml** [comment]
Ready for merging

**Comment by cdeil** [comment]
There's an issue for `input/data/2014/2014A%26A...563A..91A/tev-000016-sed.ecsv`. There's two sets of points in one SED file.

**Comment by cdeil** [comment]
I've checked that `input/data/2000/2000AIPC..515..113P/tev-000049-sed.ecsv` here matches what I computed in https://gist.github.com/cdeil/2942d2f0241d7ff8d25db8008e0fda03
So I won't review the code here: spectral point computation should be OK.

As far as I can see, the only remaining issue / TODO here is to separate out the "Low" and "High" state of `input/data/2014/2014A%26A...563A..91A/tev-000016-sed.ecsv` into two SED files:
```
$ grep '563A..91A'  other_data_collections/2015ApJ...812...60B/BiteauWilliams2015_AllData_ASDC_v2016_12_20.ecsv 
49.1793 | 41.3248 | 4.8118e+25 | 0 | 5.88161e-12 | 1.63695e-12 | 1.63695e-12 | 55120 | 55242 |  | High | IC310 | MAGIC | 2014A&A...563A..91A
49.1793 | 41.3248 | 1.10986e+26 | 0 | 6.61594e-12 | 1.17467e-12 | 1.17467e-12 | 55120 | 55242 |  | High | IC310 | MAGIC | 2014A&A...563A..91A
49.1793 | 41.3248 | 2.56307e+26 | 0 | 6.21071e-12 | 1.26734e-12 | 1.26734e-12 | 55120 | 55242 |  | High | IC310 | MAGIC | 2014A&A...563A..91A
49.1793 | 41.3248 | 5.92407e+26 | 0 | 7.55901e-12 | 1.77916e-12 | 1.77916e-12 | 55120 | 55242 |  | High | IC310 | MAGIC | 2014A&A...563A..91A
49.1793 | 41.3248 | 1.36834e+27 | 0 | 7.28581e-12 | 2.45768e-12 | 2.45768e-12 | 55120 | 55242 |  | High | IC310 | MAGIC | 2014A&A...563A..91A
49.1793 | 41.3248 | 4.8118e+25 | 0 | 8.12132e-13 | 4.73321e-13 | 4.73321e-13 | 55120 | 55242 |  | Low | IC310 | MAGIC | 2014A&A...563A..91A
49.1793 | 41.3248 | 1.10986e+26 | 0 | 7.72985e-13 | 2.63963e-13 | 2.63963e-13 | 55120 | 55242 |  | Low | IC310 | MAGIC | 2014A&A...563A..91A
49.1793 | 41.3248 | 2.56307e+26 | 0 | 1.05492e-12 | 3.22237e-13 | 3.22237e-13 | 55120 | 55242 |  | Low | IC310 | MAGIC | 2014A&A...563A..91A
49.1793 | 41.3248 | 5.92407e+26 | 0 | 1.20213e-12 | 4.0584e-13 | 4.0584e-13 | 55120 | 55242 |  | Low | IC310 | MAGIC | 2014A&A...563A..91A
49.1793 | 41.3248 | 1.36834e+27 | 0 | 8.00413e-13 | 4.25861e-13 | 4.25861e-13 | 55120 | 55242 |  | Low | IC310 | MAGIC | 2014A&A...563A..91A
```
@pdeiml - Maybe you can check if this is a scripting bug that also affects other datasets.

**Comment by pdeiml** [comment]
You are right.
In Biteau's catalog this SED is divided in two regimes "High" and "Low" but the measurement time is the same for both. How is this possible?
They can not measure two different fluxes for the same energy and at the time time, can they?

**Comment by cdeil** [comment]
As discussed on Slack just now: splitting of SEDs should be first by "note" and then by "mjd".
There are cases (e.g. `2002A&A...393...89A`) where "note" isn't filled, i.e. the splitting by "mjd" has to happen there.

**Comment by pdeiml** [comment]
I added the separation by "note" and I reordered the commits which I did last week.

Now, it should be ready for merging :-)

**Comment by pdeiml** [comment]
Ready for merging :-)

**Comment by cdeil** [comment]
I've checked `2000AIPC..515..113P` and `2014A%26A...563A..91A` again, and unless I made a mistake everything is as it should be now.

@pdeiml - Thanks for this large addition to gamma-cat!!!

**Comment by cdeil** [comment]
I've updated the auto-generated files in the output folder: 04cbd4757f4ee67c815982bac461b1dde08c8d32 and f0dfddf504b4ca450cab3e6d6339173fba3c84b9

**Comment by cdeil** [comment]
And I've re-run the scripts in gamma-cat-status:
* https://github.com/cdeil/gamma-cat-status/commit/c8988122772ed9abfda7c9882dc3978590886eac
* https://github.com/cdeil/gamma-cat-status/commit/09df17251e94400f52d8d6b51d69b22c46774a3a

Not all of the changes in those diffs are due to this Biteau PR, but some are, so I thought I'd mention it here for reference.

**Pull request merged**